lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs

```diff
@@ -112,64 +114,89 @@ random
     :: forall m e. MonadRandom m
     => CoinSelectionOptions e
     -> NonEmpty TxOut
+    -> Quantity "lovelace" Word64
     -> UTxO
     -> ExceptT (ErrCoinSelection e) m (CoinSelection, UTxO)
-random opt outs utxo = do
+random opt outs (Quantity totalWithdraw) utxo = do
     let descending = NE.toList . NE.sortBy (flip $ comparing coin)
     let nOuts = fromIntegral $ NE.length outs
     let maxN = fromIntegral $ maximumNumberOfInputs opt nOuts
+    let totalOut = sum (getCoin . coin <$> outs)
     randomMaybe <- lift $ runMaybeT $
-        foldM makeSelection (maxN, utxo, []) (descending outs)
+        foldM
+            (\acc out ->
+                let
+                    withdraw = totalWithdraw
+                        `proportionallyTo` (getCoin (coin out) % totalOut)
+                in
+                    makeSelection (Quantity withdraw) acc out
+            )
+            (maxN, utxo, [])
+            (descending outs)
     case randomMaybe of
         Just (maxN', utxo', res) -> do
             (_, sel, remUtxo) <- lift $
                 foldM improveTxOut (maxN', mempty, utxo') (reverse res)
-            guard sel $> (sel, remUtxo)
+            -- NOTE re-assigning total withdrawal to cope with potential
+            -- rounding issues.
+            guard sel $> (sel { withdrawal = totalWithdraw }, remUtxo)
```

Comment:
`sel` here is constructed by appending multiple selection together from left to right, using the `Monoid` instance on `CoinSelection`. So in the end, the total withdrawal that we get is the sum of all withdrawals that were assigned to each output, but because `proportionallyTo` rounds **down**, the final sum may end up being slightly less than the total (an error of 1 per output at most) yet, we need it to be _precisely equal_ to the reward's balance without what the Ledger will reject the withdrawal. 

Thus, this is reassigned here, and the possible exceeding will be balanced out during fee balancing anyway.

Reply:
if we transfer withdraw to one output or new one we would not have this problem;-) But, when used `proportionallyTo` indeed we need to take care of this... So what happens in detail when : 
1. `proportionallyTo` rounding does consume all withdraw
2. `proportionallyTo` rounding does consume not all withdraw?

Reply:
> To which changes this additional 1s go

Most likely, they'll be used to partially cover fee in the fee balancing algorithm here. What will happen in case of rounding issues is that the transaction will be _slightly_ unbalanced (positively) and have a few lovelace extra on the input side assigned to "nothing". Following the fee balancing algorithm, it'll start by consuming that delta, and then, starts to deplete (proportionally again) change outputs. So in the end, the little exceeding isn't lost but goes directly for fee; which is roughly similar to assigning it to any of the change output, since they'll be depleted for fees regardless.